File: app/src/main/java/com/zest/android/category/CategoryViewModel.java
package com.zest.android.category;

import android.databinding.BaseObservable;
import android.databinding.Bindable;
import android.databinding.ObservableBoolean;
import android.databinding.ObservableField;

import com.zest.android.data.Category;
import com.zest.android.data.source.CategoryRepository;

import java.util.List;

public class CategoryViewModel extends BaseObservable {

    private final ObservableField<Category> model = new ObservableField<>();
    private ObservableBoolean isLoading = new ObservableBoolean(false);
    private CategoryRepository categoryRepository;
    private OnCategoryFragmentInteractionListener listener;


    public CategoryViewModel(Category category, OnCategoryFragmentInteractionListener listener) {
        this.listener = listener;
        this.model.set(category);
    }

    public CategoryViewModel(CategoryRepository categoryRepository, CategoryFragment listener) {
        this.categoryRepository = categoryRepository;
        this.listener = listener;
    }

    @Bindable
    public String getTitle() {
        return model.get().getTitle();
    }

    @Bindable
    public String getImage() {
        return model.get().getImage();
    }

    @Bindable
    public boolean getLoading() {
        return this.isLoading.get();
    }

    public void setLoading(boolean isLoaded) {
        this.isLoading.set(isLoaded);
        notifyChange();
    }

    public void onCategoryClick() {
        listener.showSubCategories(model.get());
    }

    public void loadCategories() {
        setLoading(true);
        categoryRepository.loadRootCategories(new CategoriesCallbackImp());
    }

    /**
     * To make an interaction between {@link CategoryViewModel} and {@link CategoryRepository}
     */
    public interface OnCategoriesCallback {

        void loadData(List<Category> categories);
    }

    public class CategoriesCallbackImp implements OnCategoriesCallback {

        public void loadData(List<Category> categories) {
            setLoading(false);
            listener.setResult(categories);
        }
    }
}

File: app/src/main/java/com/zest/android/category/CategoryViewModel.kt
package com.zest.android.category

import android.databinding.BaseObservable
import android.databinding.Bindable
import android.databinding.ObservableBoolean
import android.databinding.ObservableField
import com.zest.android.data.Category
import com.zest.android.data.source.CategoryRepository

class CategoryViewModel : BaseObservable {

    private lateinit var categoryRepository: CategoryRepository
    private val categoryData = ObservableField<Category>()
    private val isLoading = ObservableBoolean(false)
    private val listener: OnCategoryFragmentInteractionListener

    constructor(category: Category, listener: OnCategoryFragmentInteractionListener) {
        this.listener = listener
        this.categoryData.set(category)
    }

    constructor(categoryRepository: CategoryRepository, listener: OnCategoryFragmentInteractionListener) {
        this.categoryRepository = categoryRepository
        this.listener = listener
    }

    @Bindable
    fun getTitle(): String {
        return categoryData.get()?.title ?: ""
    }

    @Bindable
    fun getImage(): String {
        return categoryData.get()?.image ?: ""
    }

    @Bindable
    fun getLoading(): Boolean {
        return this.isLoading.get()
    }

    fun setLoading(isLoaded: Boolean) {
        this.isLoading.set(isLoaded)
        notifyChange()
    }

    fun onCategoryClick() {
        categoryData.get()?.let {
            listener.showSubCategories(it)
        }
    }

    fun loadCategories() {
        setLoading(true)
        categoryRepository.loadRootCategories(CategoriesCallbackImp())
    }

    /**
     * To make an interaction between [CategoryViewModel] and [CategoryRepository]
     */
    interface OnCategoriesCallback {

        fun loadData(categories: List<Category>)
    }

    inner class CategoriesCallbackImp : OnCategoriesCallback {

        override fun loadData(categories: List<Category>) {
            setLoading(false)
            listener.setResult(categories)
        }
    }

}